text/1008-tls-and-alpn.rst

```diff
@@ -0,0 +1,458 @@
+binary protocol.
+
+
+Client Certificate
+==================
+
+Supporting client certificate authentication is a nice-to-have feature
+in this RFC, as implementing a proper client certificate authentication
+system can be complicated - if we also issue the client certificates,
+we'd probably reconsider the CA idea below. In this section, we're only
+discussing the feasibility.
+
+First of all, we'd want to add a new Auth method ``Certificate`` beyond
+the other two methods ``Trust`` and ``SCRAM``. The ``Certificate``
+``Auth`` entry tells the EdgeDB server which users are allowed to
+authenticate themselves using a client certificate.
+
+Then we would need a new config type to store the trusted client
+certificate fingerprints. The idea is to let the CLI generate the client
+certificates using a local CA, and the CLI tells the server to trust the
+generated certificates.
+
+The certificate should contain the authorized database role in CN or an
+X.509 extension, and that role must match the requested login user
+during authentication. As the server may support several different Auth
+methods at the same time with a customizable priority, a client
+certificate is not mandatory in TLS. But if provided and if the server
+is configured with ``Certificate`` Auth, then the client certificate
+will be used as one authentication attempt.
+
+On the client side, user may use the CLI to generate a client
+certificate (and its corresponding private) for a particular database
+role in a certain EdgeDB instance, and use the two files to establish a
+connection to that EdgeDB server. The private key passphrase - if set -
+must be securely provided through either environment variables, or API
+parameters (following Python ``SSLContext.load_cert_chain()`` style).
+We may be able to place the client certificate in the
+``credentials.json`` file so that the user don't have to bother dealing
+with the certificates any more. And we could likely skip the passphrase
+for development client certificates.
+
+
+Backwards Compatibility
+=======================
+
+While TLS will be enforced by default, compatible mode is still
+available for the server before EdgeDB 1.0, but it is only for the
+EdgeDB developers and should not be enabled by the users.
+
++------------+----------------+----------------+---------------------------+
+|            | Old Server     | New Server     | New Server in Compat Mode |
++============+================+================+===========================+
+| Old Client | Accessible     | Friendly Error | Accessible                |
++------------+----------------+----------------+---------------------------+
+| New Client | Friendly Error | Accessible     | Friendly Error            |
++------------+----------------+----------------+---------------------------+
+
+The EdgeDB development server (``edb server``) will provide a hidden
+option ``--no-tls`` to run the server in non-TLS compatible mode (just
```

Comment:
Can we have `--allow-non-tls` rather than `--no-tls` ? I think we should be able to detect whether the handshake was TLS one or binary protocol.

Reply:
Yeah sure thing - a real compatible dev server could be more handy.